File: 0430/0430.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""
class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        if not head: return head
        dummy = Node(None, None, head, None)
        self.dfs(dummy, head)
        dummy.next.prev = None
        return dummy.next
    def dfs(self, pre, cur):
        if not cur: return pre
        cur.prev = pre
        pre.next = cur
        tmpNext = cur.next
        tail = self.dfs(cur, cur.child)
        cur.child = None
        return self.dfs(tail, tmpNext)

File: notes/0430/0430.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        if (!head) return head;
        
        for(Node* h=head; h; h = h->next) {
            if(h->child) {
                Node* nxt = h->next;
                h->next = h->child;
                h->child = NULL;
                h->next->prev = h;
                Node* p = h->next;
                while(p->next) p = p->next;
                p->next = nxt;
                if (nxt) nxt->prev = p;
            }
        }
        return head;
    }
};
